### PR TITLE
config: export errors

### PR DESCRIPTION
**Pull request by pysel**
---

Contributes to: #1140 
#### Changes

- Exports errors in `config`. More context: #1140 
- Adds a package to `types` specifically for errors

### CODE REVIEW

File: config/errors.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package config
+
+import (
+	"errors"
+	"fmt"
+)
+
+var (
+	// ErrNotEnoughRpcServers is returned if the number of rpc servers is less than two
+	ErrNotEnoughRPCServers = errors.New("at least two rpc_servers entries is required")
```
Comment (sergio-mena) [inline config/errors.go:10]:
```suggestion
	ErrNotEnoughRPCServers = errors.New("at least two rpc_servers entries are required")
```

File: config/config.go [hunk 2]
```diff
@@ -881,12 +887,12 @@ func TestBlockSyncConfig() *BlockSyncConfig {
 // ValidateBasic performs basic validation.
 func (cfg *BlockSyncConfig) ValidateBasic() error {
 	switch cfg.Version {
-	case "v0":
+	case v0:
 		return nil
-	case "v1", "v2":
-		return fmt.Errorf("blocksync version %s has been deprecated. Please use v0 instead", cfg.Version)
+	case v1, v2:
+		return ErrDeprecatedBlocksyncVersion{Version: cfg.Version, Allowed: AllowedBlockSyncConfigVersions}
```
Comment (sergio-mena) [inline config/config.go:893]: You are defining `AllowedBlockSyncConfigVersions` (for the error itself), but the _switch_ doesn't use it. So we will have to be careful in the future when modifying the blocksync versions to make sure that what shows in the _switch_ is consistent with the definition of `AllowedBlockSyncConfigVersions`.
My suggestion,

* either you get rid of `AllowedBlockSyncConfigVersions` and use `v0` directly hard-coded in line 893 (anyway it is hard-coded in the _case_ in line 890, isn't it?),
* or you modify the switch to use the information contained in `AllowedBlockSyncConfigVersions`

I'd personally lean toward the first (simpler) option.

File: config/errors.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
+package config
+
+import (
+	"errors"
+	"fmt"
+)
+
+var (
+	// ErrNotEnoughRpcServers is returned if the number of rpc servers is less than two
```
Comment (sergio-mena) [inline config/errors.go:9]: this is not needed, the error message is self explanatoriy

File: config/errors.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	ErrNotEnoughRPCServers = errors.New("at least two rpc_servers entries is required")
+
+	// ErrEmptyRpcServerEntry is returned when an empty string corresponding to RPC entry is found during validate basic
+	ErrEmptyRPCServerEntry = errors.New("found empty rpc_servers entry")
+
+	// ErrInsufficientDiscoveryTime is returned when snapshot discovery time is less than 5 seconds
```
Comment (sergio-mena) [inline config/errors.go:15]: Not needed

File: config/errors.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	ErrInsufficientDiscoveryTime = errors.New("discovery time must be at least five seconds")
```
Comment (sergio-mena) [inline config/errors.go:16]:
```suggestion
	ErrInsufficientDiscoveryTime = errors.New("snapshot discovery time must be at least five seconds")
```

File: config/errors.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+	// ErrInsufficientChunkRequestTimeout is returned when timeout for re-requesting a chunk is less than 5 seconds
```
Comment (sergio-mena) [inline config/errors.go:18]: > timeout for re-requesting a chunk is less than 5 seconds

I would use that as the message for the error and get rid of this comment

File: config/errors.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+	ErrInsufficientChunkRequestTimeout = errors.New("chunk_request_timeout must be at least 5 seconds")
+
+	ErrSubscriptionBufferSizeInvalid = fmt.Errorf("experimental_subscription_buffer_size must be >= %d", minSubscriptionBufferSize)
+
+	// ErrUnknownLogFormat is returned if provided log_format is invalid during validate basic
```
Comment (sergio-mena) [inline config/errors.go:23]: Not needed

File: config/config.go (hunk 2 above)
Comment (pysel) [reply config/config.go:893]: makes sense, thanks for advice! gonna lean towards the first one as well

File: config/errors.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
+package config
+
+import (
+	"errors"
+	"fmt"
+)
+
+var (
+	// ErrEmptyRpcServerEntry is returned when an empty string corresponding to RPC entry is found during validate basic
```
Comment (pysel) [inline config/errors.go:9]:
```suggestion
```

File: config/errors.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+	"fmt"
+)
+
+var (
+	ErrEmptyRPCServerEntry             = errors.New("found empty rpc_servers entry")
+	ErrNotEnoughRPCServers             = errors.New("at least two rpc_servers entries are required")
+	ErrInsufficientDiscoveryTime       = errors.New("snapshot discovery time must be at least five seconds")
+	ErrInsufficientChunkRequestTimeout = errors.New("timeout for re-requesting a chunk (chunk_request_timeout) is less than 5 seconds")
+	ErrSubscriptionBufferSizeInvalid   = fmt.Errorf("experimental_subscription_buffer_size must be >= %d", minSubscriptionBufferSize)
+	ErrUnknownLogFormat                = errors.New("unknown log_format (must be 'plain' or 'json')")
```
Comment (pysel) [inline config/errors.go:14]:
```suggestion
	ErrUnknownLogFormat                = errors.New("unknown log_format (must be 'plain' or 'json')")
	ErrSubscriptionBufferSizeInvalid   = fmt.Errorf("experimental_subscription_buffer_size must be >= %d", minSubscriptionBufferSize)
```

**Comment by pysel** [comment]
hey @sergio-mena! thanks for review, addressed, lmk if it is good to merge now please

**Review by sergio-mena — APPROVED** [review]
LGTM
Thanks @pysel for this PR! Much appreciated!

**Pull request merged**